File: AAAA_TOP100/跳跃游戏2.cpp

#include<iostream>
#include<bits/stdc++.h>

using namespace std;

class Solution {
public:
    int jump(vector<int>& nums) {

        int step = 0;
        int n = nums.size(), last = 0, cur = 0;
        for(int i=0;i<n-1;i++) {
            cur = max(cur, i+nums[i]);
            if(i==last) {
                last = cur;
                ++step;
                if(cur>=n-1) {
                    break;
                }
            }
        }

        return step;

    }
};

int main() {
    vector<int> nums = {2,3,1,1,4};
    Solution s2;
    cout<<s2.jump(nums)<<endl;

    return 0;
}

File: AAAA_TOP200/跳跃游戏2.java


public class Jump {

    public static int jump(int[] nums) {
        int end = 0;
        int maxPosition = 0; 
        int steps = 0;
        for(int i = 0; i < nums.length - 1; i++){
            //找能跳的最远的
            maxPosition = Math.max(maxPosition, nums[i] + i); 
            if( i == end){ //遇到边界，就更新边界，并且步数加一
                end = maxPosition;
                steps++;
            }
        }
        return steps;
    }

    public static void main(String[] args) {
        int[] nums = {2,3,1,1,4};
        System.out.println(jump(nums));

    }
}



class Solution {
    public boolean canJump(int[] nums) {
        if (nums.length == 1) {
            return true;
        }
        //覆盖范围, 初始覆盖范围应该是0，因为下面的迭代是从下标0开始的
        int coverRange = 0;
        //在覆盖范围内更新最大的覆盖范围
        for (int i = 0; i <= coverRange; i++) {
            coverRange = Math.max(coverRange, i + nums[i]);
            if (coverRange >= nums.length - 1) {
                return true;
            }
        }
        return false;
    }
}




